Subject: [PATCH] my patch 1
---
Index: src/main/java/com/example/finalproject/entity/enums/Role.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/com/example/finalproject/entity/enums/Role.java b/src/main/java/com/example/finalproject/entity/enums/Role.java
new file mode 100644
--- /dev/null	(date 1718790736829)
+++ b/src/main/java/com/example/finalproject/entity/enums/Role.java	(date 1718790736829)
@@ -0,0 +1,6 @@
+package com.example.finalproject.entity.enums;
+
+public enum Role {
+    CLIENT,
+    ADMINISTRATOR
+}
Index: pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pom.xml b/pom.xml
--- a/pom.xml	(revision 29f51bc3776c6ff7135f40e644cbca7d66aac3cd)
+++ b/pom.xml	(date 1718789116516)
@@ -38,6 +38,10 @@
             <artifactId>modelmapper</artifactId>
             <version>3.2.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
     </dependencies>
 
     <build>
Index: src/main/java/com/example/finalproject/entity/User.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/com/example/finalproject/entity/User.java b/src/main/java/com/example/finalproject/entity/User.java
new file mode 100644
--- /dev/null	(date 1718792112998)
+++ b/src/main/java/com/example/finalproject/entity/User.java	(date 1718792112998)
@@ -0,0 +1,39 @@
+package com.example.finalproject.entity;
+
+import com.example.finalproject.entity.enums.Role;
+import jakarta.persistence.*;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Entity
+@Table(name = "users")
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "UserID")
+    private long UserID;
+
+    @Column(name = "Name")
+    private String name;
+
+    @Column(name = "Email")
+    private String email;
+
+    @Column(name = "PhoneNumber")
+    private String phoneNumber;
+
+    @Column(name = "PasswordHash")
+    private String passwordHash;
+
+    @Enumerated(EnumType.STRING)
+    @Column(name = "Role")
+    private Role role;
+
+
+}
